### ISSUE TITLE
failure with kernel-install script

### ISSUE THREAD

**Issue by cvlc12** [systemd/mkosi#3048]
### mkosi commit the issue has been seen with

24.3

### Used host distribution

Arch

### Used target distribution

initrd

### Linux kernel version used

6.6.52-1-lts

### CPU architectures issue was seen on

x86_64

### Unexpected behaviour you saw

Crash when running mkosi kernel-install plugin. Maybe linked to pacman update ?

### Used mkosi config

```ini
none
```

### mkosi output

```
# kernel-install add 6.6.52-1-lts /lib/modules/6.6.52-1-lts/vmlinuz --verbose
layout=uki set via config
INITRD_GENERATOR (mkosi-initrd) set via config.
Loaded config.
Loading /etc/machine-info…
MACHINE_ID=9821320080000801887cbe1 set via /etc/machine-id.
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Found container virtualization none.
Directory "/boot" is not the root of the file system.
Couldn't find an XBOOTLDR partition.
Using EFI System Partition at /efi as $BOOT_ROOT.
Using entry token: 9821320080000801887cbe1
File '/etc/kernel/install.d/91-sbctl.install' is a mask.
File '/usr/lib/kernel/install.d/91-sbctl.install' is masked by previous entry.
kernel version (6.6.52-1-lts) set via command line.
kernel image file (/lib/modules/6.6.52-1-lts/vmlinuz) set via command line.
Using ENTRY_DIR=/efi/9821320080000801887cbe1/6.6.52-1-lts
Using plugins: 
  /usr/lib/kernel/install.d/50-depmod.install
  /usr/lib/kernel/install.d/50-mkinitcpio.install
  /usr/lib/kernel/install.d/50-mkosi.install
  /etc/kernel/install.d/55-lsinitcpio.install
  /usr/lib/kernel/install.d/60-ukify.install
  /usr/lib/kernel/install.d/90-loaderentry.install
  /usr/lib/kernel/install.d/90-uki-copy.install
Plugin environment: 
  LC_COLLATE=C.UTF-8
  KERNEL_INSTALL_VERBOSE=1
  KERNEL_INSTALL_IMAGE_TYPE=pe
  KERNEL_INSTALL_MACHINE_ID=9821320080000801887cbe1
  KERNEL_INSTALL_ENTRY_TOKEN=9821320080000801887cbe1
  KERNEL_INSTALL_BOOT_ROOT=/efi
  KERNEL_INSTALL_LAYOUT=uki
  KERNEL_INSTALL_INITRD_GENERATOR=mkosi-initrd
  KERNEL_INSTALL_UKI_GENERATOR=
  KERNEL_INSTALL_STAGING_AREA=/tmp/kernel-install.staging.uPZxMw
Plugin arguments: add 6.6.52-1-lts /efi/9821320080000801887cbe1/6.6.52-1-lts /lib/modules/6.6.52-1-lts/vmlinuz
Successfully forked off '(sd-exec-strv)' as PID 9154.
About to execute /usr/lib/kernel/install.d/50-depmod.install add 6.6.52-1-lts /efi/9821320080000801887cbe1/6.6.52-1-lts /lib/modules/6.6.52-1-lts/vmlinuz
Successfully forked off '(direxec)' as PID 9155.
+depmod -a 6.6.52-1-lts
/usr/lib/kernel/install.d/50-depmod.install succeeded.
About to execute /usr/lib/kernel/install.d/50-mkinitcpio.install add 6.6.52-1-lts /efi/9821320080000801887cbe1/6.6.52-1-lts /lib/modules/6.6.52-1-lts/vmlinuz
Successfully forked off '(direxec)' as PID 9158.
/usr/lib/kernel/install.d/50-mkinitcpio.install succeeded.
About to execute /usr/lib/kernel/install.d/50-mkosi.install add 6.6.52-1-lts /efi/9821320080000801887cbe1/6.6.52-1-lts /lib/modules/6.6.52-1-lts/vmlinuz
Successfully forked off '(direxec)' as PID 9159.
‣ Building initrd
‣ + timedatectl show -p Timezone --value
‣ + timedatectl show -p Timezone --value
‣ + findmnt --target / --output FSTYPE --noheadings
‣ + stat --file-system --format %T /var/tmp/mkosi-workspace-hzxtwzx8
‣ + btrfs subvolume create /var/tmp/mkosi-workspace-hzxtwzx8/root
Create subvolume '/var/tmp/mkosi-workspace-hzxtwzx8/root'
‣ Copying in package manager file trees…
‣ + cp --version
‣ + cp --recursive --no-dereference --preserve=mode,links --reflink=auto --copy-contents /tmp/tmp5w2obs6t /var/tmp/mkosi-workspace-hzxtwzx8/pkgmngr --keep-directory-symlink --no-target-directory
‣ + rm -rf -- /var/tmp/mkosi-workspace-hzxtwzx8
‣ + mount --make-rslave /
‣ + mount --rbind /etc /etc --options ro,nosuid,nodev,noexec
‣ + mount --rbind /opt /opt --options ro
‣ + mount --rbind /boot /boot --options ro,nosuid,nodev,noexec
‣ + mount --rbind /efi /efi --options ro,nosuid,nodev,noexec
‣ + mount --rbind /media /media --options ro,nosuid,nodev,noexec
‣ + mount --rbind /usr /usr --options ro
‣ Building default image
‣ + stat --file-system --format %T /var/tmp/mkosi-workspace-0p_x2n12
‣ + btrfs subvolume create /var/tmp/mkosi-workspace-0p_x2n12/root
Create subvolume '/var/tmp/mkosi-workspace-0p_x2n12/root'
‣  Copying in package manager file trees…
‣ + cp --version
‣ + cp --recursive --no-dereference --preserve=mode,links --reflink=auto --copy-contents /tmp/tmp5w2obs6t /var/tmp/mkosi-workspace-0p_x2n12/pkgmngr --keep-directory-symlink --no-target-directory
‣ Acquiring lock on /var/cache/pacman
‣ Acquired lock on /var/cache/pacman
‣ Acquiring lock on /var/lib/pacman
‣ Acquired lock on /var/lib/pacman
‣ + cp --version
‣ + cp --recursive --no-dereference --preserve=mode,links --reflink=auto --copy-contents /var/cache/pacman /var/tmp/mkosi-workspace-0p_x2n12/root/var/cache/pacman --keep-directory-symlink --no-target-directory
‣ + cp --version
‣ + cp --recursive --no-dereference --preserve=mode,links --reflink=auto --copy-contents /var/lib/pacman /var/tmp/mkosi-workspace-0p_x2n12/root/var/lib/pacman --keep-directory-symlink --no-target-directory
‣  Installing Arch
‣ + pacman --root=/buildroot --logfile=/dev/null --dbpath=/var/lib/pacman --cachedir=/var/cache/pacman/mkosi --cachedir=/var/cache/pacman/pkg --hookdir=/buildroot/etc/pacman.d/hooks --arch x86_64 --color auto --noconfirm --sync --needed --assume-installed initramfs filesystem
resolving dependencies...
looking for conflicting packages...

Package (2)      New Version   Net Change

core/iana-etc    20240814-1      3.98 MiB
core/filesystem  2024.04.07-1    0.02 MiB

Total Installed Size:  4.01 MiB

:: Proceed with installation? [Y/n] 
error: failed to commit transaction (unexpected error)
Errors occurred, no packages were upgraded.
‣ "bwrap --die-with-parent --proc /proc --setenv SYSTEMD_OFFLINE 1 --unsetenv TMPDIR --dir /tmp --dir /var/tmp --unshare-ipc --dev /dev --symlink usr/bin /bin --symlink usr/bin /sbin --symlink usr/lib /lib --symlink usr/lib /lib64 --setenv PATH /scripts:/usr/local/sbin:/usr/local/bin:/usr/bin:/var/lib/flatpak/exports/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl --uid 0 --gid 0 --cap-add ALL --bind /var/tmp/mkosi-workspace-0p_x2n12/root /buildroot --bind /var/tmp/mkosi-workspace-0p_x2n12/pkgmngr/etc /etc --ro-bind /etc/ca-certificates /etc/ca-certificates --ro-bind /etc/pacman.d/gnupg /etc/pacman.d/gnupg --ro-bind /etc/pki /etc/pki --bind /etc/resolv.conf /etc/resolv.conf --ro-bind /etc/ssl /etc/ssl --ro-bind /etc/ca-certificates/extracted/tls-ca-bundle.pem /proxy.cacert --bind /var/tmp/mkosi-workspace-0p_x2n12/repository /repository --ro-bind /usr /usr --bind /var/tmp/mkosi-workspace-0p_x2n12/root/var/cache/pacman /var/cache/pacman --ro-bind /var/tmp/mkosi-workspace-0p_x2n12/repository /var/cache/pacman/mkosi --bind /var/cache/pacman/pkg /var/cache/pacman/pkg --bind /var/tmp/mkosi-workspace-0p_x2n12/root/var/lib/pacman /var/lib/pacman --bind /var/tmp/mkosi-workspace-0p_x2n12/pkgmngr/var/log /var/log --bind /var/tmp/mkosi-workspace-0p_x2n12/tmp/mkosi-var-tmp-4970ebc78f7a451c /var/tmp sh -c 'chmod 1777 /tmp && chmod 1777 /dev/shm && chmod 1777 /var/tmp && chmod 755 /etc && exec $0 "$@"' setpgid --foreground -- pacman --root=/buildroot --logfile=/dev/null --dbpath=/var/lib/pacman --cachedir=/var/cache/pacman/mkosi --cachedir=/var/cache/pacman/pkg --hookdir=/buildroot/etc/pacman.d/hooks --arch x86_64 --color auto --noconfirm --sync --needed --assume-installed initramfs filesystem" returned non-zero exit code 1.
‣ + rm -rf -- /var/tmp/mkosi-workspace-0p_x2n12
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/mkosi/run.py", line 60, in uncaught_exception_handler
    yield
  File "/usr/lib/python3.12/site-packages/mkosi/run.py", line 101, in fork_and_wait
    target(*args, **kwargs)
  File "/usr/lib/python3.12/site-packages/mkosi/__init__.py", line 4761, in run_build
    build_image(Context(args, config, workspace=workspace, resources=resources, package_dir=package_dir))
  File "/usr/lib/python3.12/site-packages/mkosi/__init__.py", line 3915, in build_image
    install_distribution(context)
  File "/usr/lib/python3.12/site-packages/mkosi/__init__.py", line 152, in install_distribution
    context.config.distribution.install(context)
  File "/usr/lib/python3.12/site-packages/mkosi/distributions/__init__.py", line 114, in install
    return self.installer().install(context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mkosi/distributions/arch.py", line 41, in install
    cls.install_packages(context, ["filesystem"], apivfs=False)
  File "/usr/lib/python3.12/site-packages/mkosi/distributions/arch.py", line 45, in install_packages
    Pacman.invoke(
  File "/usr/lib/python3.12/site-packages/mkosi/installer/pacman.py", line 167, in invoke
    return run(
           ^^^^
  File "/usr/lib/python3.12/site-packages/mkosi/run.py", line 150, in run
    with spawn(
  File "/usr/lib/python3.12/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/usr/lib/python3.12/site-packages/mkosi/run.py", line 352, in spawn
    raise subprocess.CalledProcessError(returncode, cmdline)
subprocess.CalledProcessError: Command '['pacman', '--root=/buildroot', '--logfile=/dev/null', '--dbpath=/var/lib/pacman', '--cachedir=/var/cache/pacman/mkosi', '--cachedir=/var/cache/pacman/pkg', '--hookdir=/buildroot/etc/pacman.d/hooks', '--arch', 'x86_64', '--color', 'auto', '--noconfirm', '--sync', '--needed', '--assume-installed', 'initramfs', 'filesystem']' returned non-zero exit status 1.
‣ + tput cnorm
‣ + tput smam
‣ "mkosi --force --directory '' --format cpio --output initrd --output-dir /tmp/kernel-install.staging.uPZxMw --cache-only=metadata --extra-tree /usr/lib/modules/6.6.52-1-lts:/usr/lib/modules/6.6.52-1-lts --extra-tree=/usr/lib/firmware:/usr/lib/firmware '--remove-files=/usr/lib/firmware/*-ucode' '--kernel-modules-exclude=.*' --kernel-modules-include=host --include=mkosi-initrd --debug --workspace-dir=/var/tmp --package-cache-dir=/var --include /etc/mkosi-initrd --package-manager-tree /tmp/tmp5w2obs6t" returned non-zero exit code 1.
‣ "mkosi-initrd --kernel-version 6.6.52-1-lts --format cpio --output initrd --output-dir /tmp/kernel-install.staging.uPZxMw --debug" returned non-zero exit code 1.
/usr/lib/kernel/install.d/50-mkosi.install failed with exit status 1.
/usr/lib/kernel/install.d/50-mkosi.install failed with exit status 1.
(sd-exec-strv) failed with exit status 1.
```

**Comment by DaanDeMeyer** [comment]
@cvlc12 Please reproduce with the latest git main commit mkosi has changed significantly since the last release

**Comment by DaanDeMeyer** [comment]
@cvlc12 Please give https://github.com/systemd/mkosi/pull/3055 a try

**Comment by DaanDeMeyer** [comment]
@cvlc12 Any chance you can patch mkosi to call pacman with `strace`? It's not telling us anything useful in its error message

**Comment by gdiscry** [comment]
Here is what I've found.

1. `mkosi-initrd` creates an empty `etc/pacman.conf` in the sandbox.
2. If `/etc/pacman.conf` exists on the host, `mkosi-initrd` copies its content in the sandbox.
3. When `Pacman.setup()` is called, `etc/pacman.conf` already exists in the sandbox and is kept as is.
4. Since Pacman 7, a sandbox is used with the user `alpm` to download the files (as configured by default in `pacman.conf`).
5. `pacman` cannot find the user `alpm` in `mkosi`'s own sandbox and fails.

I tried adding `"--disable-sandbox"` when calling pacman, without success.

If I remove the line `DownloadUser = alpm` in `/etc/pacman.conf`, `mkosi-initrd` works (24.3 and git main).
